{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# ChIPAID project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": []
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAANEAAADACAIAAAAV0hGGAAAAA3NCSVQICAjb4U/gAAAAGXRFWHRTb2Z0d2FyZQBnbm9tZS1zY3JlZW5zaG907wO/PgAAIABJREFUeJztnW1MG1ma758d3aWs3XRZm8hWlYS73FrbrNqYUWzTCZZpxdeMZExL4ERDSCaNIBIx0SUmEoFoQNBJBIwytyMlJlfhRUoEnc4CIwaQMsbSpHE0EabTDY5mwCsNeDUdCinBaLJUJStcWWl8P5TfcGwDSbq6xzm/L8mpc+qc5xz/fV6Kevz8QzgcBgRCQH7yQxuAeOdAmkMIDdIcQmiQ5hBC879+aAO+X/7zP//zhzbhHeUnP/lJTk7Ov/zLv/zTP/1TctYPYhAi6/nb3/4WCoWePHny3//930lZSHOI75f/+q//SrqCNIf4fvmf//mfpCtIc4jvl7/97W9JV95hzT3/+v8etxzv+sPzN6jj6y6L0Wg0Go3/+9Ttx2/Nsiwny8+tAPDk6xvO3qlHKy9e7iEUauvpcyf27wMAgJdPV56+eLrv2QuA91638v31fV+egCcTn7X6d2TL6JlfTHzYd+e0kk8v3zhu/4+KL3vUE8ftX65sLfv+L/rufPyH1NdPqx9dq+maWnn64mXOnr2EurjyZH2F+r3M/f3RkO2aezL6WesklDouNFPvPX/if/QiZ080a19F95fqZ3uU5BtUj+2jqH0A+3LezEpF5a++LHn5Ep5MfHbh0f4LFytIyMnZS8AeIvV1gOfPVp6p62/V74eXf/VP9DqbLuR8ecW6L1N/fzRkueY4/6NATnG3w3oQAwC1+mDk+tddlnNTLwAg58P6W/0nqMjlv97r+mXvg8CLnPcPFCufPXywnKM+0XbhhPo9zj/aem1ieeXps5c5hOJA5WlH5Xazx3P/qLN34uHy05d7FMUnGh3ReSgVEelCDpkDfpJSKqMGYWmuAwDk7CGUSiWAUqmARw9av3r43GrNSdPf9PZwy67e21OP/I+fPnvxEnL2Fp/ru1jy7MZx+4MDPXca1fy9j2/X1Lr28+nd1JN6iLJccxj1PvHywVcP/3rw4y39T7MmPnu6HCBP9DiI0V92+w9cuPLzh10XeqfMPZV7nzx6FNh3srttP/Z82dXb2/TLl339J5TpG3482tR0Gyt3XHSQnH/U6fylc++XbR9/P50EDHIiE226/qax5z2AZ4+mJh/l/NzRepB8LwdePof39wHsO7h/728ePXoMagoA4MnDhyt79zcqdl1ParJcc6A8edHx9LPuXxyfNFf+/ETpQRIDgExrYs4eQqE8cIDKeUIp1R/nqPd8/vgpwF4AyCEUB9VqAPV+8uXj2tujX/+87SCWulXu69u3nx449+Xpj98DAGXry0fl3V894j5WAsDKl7XGLxPa2/+GPXz+5NHklw9fKE6o30vb33T2fMzbv0dRUnJwyxfow5IDe6b+8PDJCYoEePLwgX/vgZMfYruvJyXZrjnAlBUX75gf/2Fq9Dfdv7itrP9Vd6UyjVKSeQmQuiS1fz/R+yjwFA5SKfPhaeDxi2f/caH8Qayql7D/2QsAACBKu69EF/PHt5suPN15X7by4qtW41cAADl7PyxtvVjJ15myv2ntSTcS2IclxXunHjx8UllBPnn4lX9vcf1+DODxbutJSdZrDgAA3qM+rmz+2Hygy37h/04W91fu7NiQ+WDwcpvc98sv/KoyrsmcPXv3wQsAyNlLUlTk+su9OfDamss54OhzHMjJ2UuQ7yV95sn9TWdPerAPSw7sbXI9eFJR/MDlJ8wOdaZ+7Y53Q3M8+/bvp+Dh46cAEc3l5GDw4vmLV0u+Z73itgJAGkH81f8fT3Oo94loOicH4MXzBBESCuWep8srQH28dSJM0dTrk7OHUFJpJloASOzvh2nsyQC2v8JM2F1f3fvrvYCioo1fMdP1a1eEslxzz7++3RvYe0BNkTkvn/gnbj/KUZ97P5a7V/n+nqcPRif2lxIvnoHaejBDTQAvHv3mhuvl/r0vHn5500+U9hyITS573ydznj64zdeDqa37yYOVlYozN5uanp8sP0i+x/318bN9xVYBnpOl6S/2OvYoKyoUv+ntXsnZf84ckdhr1ZOMKMs19xLLeflo1Hl75ekL2Pu+en/9rxzW+Ahh+0+cK112Olsf5BDqkxe30VxOzvNHN7t/8wwItbn1V/Xq+HqGHTzpMHf1xurZD6A8ceXKnt7e0d4Lv3kGOYTiwMkD1jc9LeyAtP19HXvI0soDNy8sF//cHB+yt9Gvf8huH5y39f4cd++XpZ/v6Xa3ZZYlIiX/+q//mph8h//eihCEn/wkWWNIc4jvl3/8x39MuoLWVsT3C0EQ//zP/5x4JcvPEIgfigz+EFk+zyF+hKD9HEJokOYQQoM0hxAapDmE0CDNIYQGaQ4hNNmvucWuYlJeOxXaSckSkiRJMq/RE70UmqiVkyVdi9+ngTsihW3J0P2H5fLDQ8HMFXlbCkiSJMnijvnMBec7CkmSJMnCFu/OzdyRDdmvObHWUllukoq2L5nfdOf3V8slW67JisorTVrp92AX3X9YTpIkSZLykp5lAABgJmrzyIIWL//9iH7oJCk/NiJruDF20SLJVCGuNJWXmxTbvLKrPTc21lC0g/d61Q13xloP4TvrzO5sCCMSmWvXEyrHtBBNbW7MdVopyto5t7bJX5h2aAiCMHbORQqsTbcbCcrWt7Txdm1z2VWEsX1u+4IzDg2hb5558xa3kP3z3I8WkVhn0Uk4Osji/CTs9/jWcRynvfP8tCfCQ8EgpjCZlOIf0s63TnZrju4vk5MkGd3PRZezwpYpb399WWGeXF5QUtvjZVLf7WnMI0mSjO7nvI0FJEmS8sP93omOw8V5cnle4eGWieXEnWLQ21NfUiiXywuKjyVlpUJdpMNZ/yINAADL3nlaYio3YH6Pj98Q+WfnOVyr2+pJFfT215cUyOXygpL6/vmI6czEMb6jCXupED3VUVtSmCeX5xWUHGsZmk/aZdFTXcdKCuTyvMKyxpEMpnKBkZbDxXlyeUHxsa4p3laY7yjmtwVdUyONZQV5hS3eJBtCE7V8sn5kfqiRH+ziYx1TdLavrZtrC8N2PUHVuDb55JLLYSQIlb6qfXxmYWHO1WlTERq7ayN6w5b1a2NlxmmjCHPnQiS54LRRBKUx2/um5xYWZgbtRooyO5ci927MtJspSl/lHJ92DTZbVYTeMb0RzsjKoI2iqoY3wuHw2qCNUtWMLw1XUSr7+EY0s2Z8i22ESmO1O4fHx4c7q/QEoW+eiazLa0uuZjNB2QbXIsa4HHpKZXb0uaanXYPtNg1B2QZX+DyXXUVQKr3V4Rx2jQ822zQEFe3jVmYcGoKgNGaHc3h8fLDZqiEos3MhOrKDVSqCUultzc7BYdfCRrINmxsrM+1WilBpbI7BmYWlhWlnlYZQVQ1nuebC4fBCpzGmuXA4vORM/GTCmzPNGkLjiG1ZkvZMm+M1cc2Fw2GXnYp/zhFVRD/KlT4rRRjbo5lLTmtUThmYazdSmuaZcHhjvEZF2YbXwit9VkrjmN4Mb7rsKsrat5RQVk9QVcNR0zeGa6jEbdlKn42K9WzJaaUoW99KbBScZsoY7YfLriI08e/DSp+V2tJQjBmHhlDFVb82XqMiNI7pzZS5r9jAf4+IqErD0c8iu9fWdIjx6HlMhIvFwLLBHTxKiSIS49FDsAQXA8fSDAAA4/X4udwiixZCoVAoFJLlK3Au4A9krkxdpMZY3/xyyOeZ59QWrRRkBoOM9Xn8If+sn5Oo87eurJhUGj1Ei2UyCQQZNlW1jM/rB4XJJIteyK++MXajWpEwBGI8ukuUyGTAsaspKwLAJdJoQanJosXW/YsJfcLiuWnAcCxWQiyTYMCg9+d25w+cAo7/h2VYgNUvjnzwRUKehOEy3yzSGdTg9vpm132szGJQAkC+qUgyMO/1KRZpLL9J/VqGswwLgIsT5CBW6tKJQwQAsKMvnVgqxSFAp1HnTkGae1vgYhxA8enNq0fjT/MwTKrIcAsAgFSrk8GEe4SlJUWmfAAA0Jp02Kh3yBsARUNaoezAmHWGAXirjxZDIY4DkegNv6Xv5tr6fSDWGtRYkGZlujj5+bJtn0UrDFp81eP2YzqTlr8iMpi0MOtxr0t0OlnmmzMYo4CAZ5aOXljuP1xQ3LGLvyjEYNeD0QNvyOeZZ3GFdrvv0TZkueZCweVAMATABPw0AxBilgMBDoCh/TQTzWZi6VBwORDkAILLi8EQxC4zwcAyEwJggot0ECAUDCwnZkfTyupz5dLZjuON/VMer2dqpKf+cO1Q5DHISEt9y8hyShNFaoMa4zgwmAxRfYoNJi2wHK4riq+sIWaZDnIALM03HmKWA0EOuCCfDjF0YJXlgOU7CsrqJotktqu+cWjK45noqT99xSc1lWv5Pq8zHIQYmmZCfJpmACAYSSfDebtqG/tHpiaGumobv1jNLa+ziAEgFFwMsBxw64HF6GY42QaGDgQ44NjVAF8zQwcCLHBsdp9bV/qsFBFB45hZ6bPFkir79GZitsoxvdBpjqYIfftMeNqhiiYpW9/KjEMTS1qdS5G/GkTSkRPiyrTTbtWrKIpS6a12p2uFP+LNNGd6nh85Sq8kXFroTLqUZNvmkjMxPbcxXhXrCVUzzj8YWhpvrzJrVBFbZtZipkQLmp1LmzPN+sT0VstmHBrK1jc93GzTqyhKZbQ1jy/xPZprN8baj5yHk2zYGK+JJ6uGN+Jp5A+BEJosX1sRP0KQ5hBCgzSHEBqkOYTQIM0hhAZpDiE0SHMIoUGaQwgN0hxCaJDmEEKDNIcQGqQ5hNAgzSGEBmkOITRIcwihQZpDCA3SHEJokOYQQoM0hxAapDmE0CDNIYQGaQ4hNEhzCKFBmkMIDdIcQmiQ5hBCgzSHEBqkOYTQIM0hhAZpDiE0SHMIoUGaQwgN0hxCaJDmEEKDNIcQGqQ5hNAgzb0Kw0fkk9dOpAl4uG0BRDIheqLjWHGBXC7PK353NOdtKYzEfH7LQlnuKSFJkiz+EURQ/9GyPFR/duB+YJ3jOPbdmefm3Z5VAADO5/ag2ektw8wPdbQ0pg27AozP4+cAQP3pnT/8cexdifc1P8lLDoD1ur1MRenbC/0s1dW1tgZBrHurobX+rgh6R4a+8HG54rrU+ZFAe7i61KSUvisx5hbdngAAYDjGsaxn0hsqLd02ENdOERuOnjG8rcr+7giFQLTLoXw31tZljycAgKnrGiwSANbrnt0a2Sro7W88XJgXiWk+4ntl8c1UYOt+bqpeTpLyso6R/sbD/J65sKxxaJ5Jqqu4IE8uzyssq+8ZGek4VlxQ0OhJbjNET3XVlvGxzQtLjrX0e4MAsNxTJidJsrBjPlrO21JAkqS8rGeZTxVG4rdHbpYXlBzriIdp560t6fJ4eurLivkeHW7ZYiAAszgSCasuzyssq+2YWIyNV4jv4OH+RT7ieXHb/6vP+7jbxwHA6sAnJEnmJXXF21Lw07P3OQBgR4+TJCl/J+a5Zbc7AAAKU3m1mr4+ObnumZgNmUyRryfjaTxSOxrg5//1wP0vPru/9fZtC7wC5xs468PwXIVCwvlXfaPna1n83q0KKUDI23Hk+ECAk6gtFgMz757sPjuJ5RZZKixJwQIZz/nDJ0dXATBJrgTYwP0vrkhLTxl2tn5zs58dmQUMVygUEtrvvz9w+pPF4N3fnorFH/ZfP34cAM9VKyS0PzD7xfkjPnrsbptOBACMt+PI8QF/ZD3kVn3ugdMer+/O2CVDbEPC+a4fP7K+zgGosRADGAYcBwAYhgGWFGkzFGRj+YBhGODvwjxHe9x+DkBhsOSLTSYtBrDunfJFg/F5u1pGAxxg6k9vfvOXJ3/55k6ddsuobVsgJZJDF+/+8dt79769d7kIB1j3DEzQABAc/XwowIG64c7dW729v/3d5UM4ACfSNVwq3Ro1M+QZmlwFwA9d/eZP3377pz9/c/dqU7IsM4IXtd7944N79759cOdTBQA7e6VjJJiQj6nrxr759t69b7+526DFgPMPdAws891tHPBzgBc1jH3zlyd/+WasoQgHzj/Q2OVNWBy4dVZW3nrjztjV0//n3//8+yYtBgC51WPffffnP10zJRoiquj90zeXD2EAgFfe+u677/70DmguIrncIosOQGwq1WIAqx63DwAAQrMj7lUAyK2+dKlUJgKRzNR6KTGG+LYFUoKpK47miwAAZJajagyAC/gCIQD/vJ8DyNWa+EypqVSNAdDzPjqpgmh4a47hA1xKdUdPle4imGtu+bm6aBvnGg5hAOzsiDsuOtzS1MRPW2Jd0zmLBIDzTbqXgeG1jmmbLrcZZCIQyQxtlxu0GMDq5JAnLjqs6NKd3jMVJsMOItQmk/2aC3omfByAxFCqBYjElAdY9bjnAQDoRT8LABKDSRsdO5HapMVjt29bYDtwXAoAwHHxHVM8XjoX4v//SuBUscFShAOws5/97GfHuiYWd/l4B8s3xMwFqdagAAAusBiI5asN+bGFUqQ2KDAACPj8ocDsPAsAMoMhtg6D0qCTAQDrn/UntIDtfAiSyHrNBT1TPg4A1kePf0CSJEl+9NksBwABj3sRANh1jgMAsVQS/7qKcHF8PLctsHM4ALVBiwOse0amGAAILU5M+jkAWf6r8ailR3vHLperceAC96+f/tlHZR1TyXNhBjBMnDD9iDAMAwCWjQcF3iIZDOf1x3Asy7AAgOGSxGdJeC6OAQDDcPA2yHbNMR73bOqRCngmFwFwCcYPJ5uQFZ19AHZQYFdIy881aDFYHT35UWFJ8UefdPs4kFiaGnQpyorzq3vvffP7Gw2HcjFgfQOnG4d2rDqOCSbMjMw6ywEAjseFyCb2h40UFmM4LpYCAMeuJ06s7CrLf++238fuhCzXHONxezkATNtwZyzO5XIJAPg9nmWQ5atxAGB98/FH6Mv8csqzbYHdIdJVVJskABiwAZqTaMtbb/7uRkX606g4v6Lt3+9dteAAnG/Ky4BIjAEAcExGCzi/xxtTDe310gCAKfIVifmxzV3Qy/+VQKFVixRFapy/JaG73nkaAHC1Nh6q/U3Ibs0xXreXBcC05dUmQ5zKoyYJAATc7mVR0VFLLgDnH+ge4WcReuLKgC8+jW1bYHeEpro63Ougrr71zXff/fnb3/WeKU29Cff2NPZM8aHuQ0E+4j0mwjCQKBQ4AKx7J/g/4dFTI7Mp5Lfu7j4/sswABOd7Wq77OAC86Kglrm3uftf5/nmGfxT3+X0WAFOXW5QgNlWX5wJwvuvne7zBEISC3p7z130cgOJotSndcUHET4Dr897FV3amr5LVz+dC3kkvC4CpTaYt+yVRkcWAj06yfreHPnOq7deVs7WjAffZ4sIBtZjx+1e3CEpk2KbA7hBLcQDWP3DkpyO5aoPBYLJYLCadNPnTDAY8k93do90YhgOwHAcAuZZqkwhE2qNHFaPXA6ujtR95ZVKODqyntIYLTJ79ePJsNIkXNbWVb5lO192fffJvn0VTmLr6Up2S7+61Ot/xAf9s95Gfdsez6y43GdKeUCX5agn4Vjlf988+6M6tG/v2UqY/zGTzPBeadXvXAUBhMCm35ogMFgMOwPndHhrEpmtjdy6Wa3Oxdb+fBnV5Q4MlN7H0tgV2gUhbWV2EA2AYsKs+9+j18yc/+eijw12e4NZyuOlM66eH1LkYsBzgudryhpu/vWwSA4BI13TnRt0hhQS49WBIaqq72mp55USDqT+9WGdR5+IYhkkUhz69OnbrVH6iZHLLW1vLtbk4BhiuKKq8ODZ2KSopseHS2N2r/N2A4blaS92Nu4kPhFP0ynDuWsMhBY4BACbNlWwzCGGEgGxMO/QEoakZXNoMb64sTA932s0qgiAIjd21+VZamGnWEwRBWftWUucvOc0EQRDGzoW30tzrkM3z3I8Pxt0/uQq5FQ3VShGIZPmmo229Y5ctGMB6IBDY/v7sAGlOUDARAKx7JxMOjT5vgAPAJNJ35lWorD5D/OgQm6orFZ4v/ANHPhrJlcnEwND0KssBrq07V/7OaA7Nc4IiNv16bOxynUUrw9Zpv59mccWhytabd8f4VzreDf4hHA7/0DYg3i3QPIcQGqQ5hNAgzSGEBmkOITRIcwihQZpDCA3SHEJokOYQQoM0hxAapDmE0CDNIYQGaQ4hNEhzCKFBmkMIDdIcQmiQ5hBCgzSHEBqkOYTQIM0hhAZpDiE0SHMIoUGaQwgN0hxCaJDmEEKDNIcQGqQ5hNAgzSGEBmkOITRIcwihQZpDCA3SHEJokOYQQoM095YITdXKSZIk5WX9u4jLJQC7MkyQXmS35vgQzqko6Vl8u02JTJfv3vxU/aYRseihw/IthsrzCooP1/d4XlsCuzJMZLr6+7GGHcSnfROyW3OGS/d+f7EIg9zyy4nhviyStzKoU/XyhBjlImm+QbddNI5tkVXeenCzUhGPUHbrapsFm79Se7xjfgdhjVKxK8NEYqVa8dpRMndGlv9uukiqkEkB1mVqgyEWOxCfl8KPa/1LQCSWKRQYgFStNURCz5hMOP3Tk1Mjvku6TCGN/m7I7nkuNfl1Y38cq8sHAICgt7++rDBPLs8rKKntmKL5uYSZOCYnSZLMqx/x9Bwrzss7NpIUG2mqPo88OcnB6sAnpLysZzkhiwtMdBwuzpPL84qPbQ26GqKnuo6VFPCNde08HqtIjOORgPYAEPQOtdSWFObJ5fK8wrL6/ngAQ6A9XZGcguKy+q6J5S1TY3rD0rHYVRJd4msn+LpCyyMtfCWFZfU90TgX8x3FJEmS8pKuqZHGsoK8whavt7GAJElSfrjfG2228HDLxHIo+2MvbbrsFGHsnEuZN9dppih9jXN8etrV5zCrCL3dtRYOh8ObG0vT7WaCoDTmms6+weGZ5EhGGysLc31VFKG3D88tLK3xcZM2hqsoglIZbc19467xPodZRVC2weitay67nqKM9r7xaVef3UhR5s65lPGWlpxmgqoaXtvc3Nzc3NxYmXHaNKpY4Wm7itDYmgdd09OuwWarilBVDfNNzHUaKcpo73NNT7uG26v0BFUzvhPDkrs2XEVFYjdtLg1WqQjKaB+cXtoIh8PhleEqDaGyOgZdrnFnlZ5S2fqW+KFcWxisUhGUSm9rdg4OuxY2whsrC04bRVAas71vem5hYWbQbqQos3PpHdFc+9xmlHjW2niNijC2z0QvrQzaKMocDYS10mejCHOmsFguO0Xo2xPUvDFcxbcVSc+1GwmqangjHA6HwwudRoIyO5ciZs21GwmNYzqV6CIhuRLQ2PoWoiVnOq02Z6zRhU4jQdWMb4TD4ZU+K0XZBteitfRVGe3jG9sblkxcc5tz7Vu+GpvTDg2hqhqONLHhsmsoY2wAZhwaQlUznlipy04R+uatA2wbfDfW1sDAJx9EIQtavPxVv8fL5hZZYiEiZaYKLQTmffFlFMNlu41Og+G5scCYMnUuFo0ITXu9NCjKTbJQKBQKhUChVWBsYDHdCocVtY7d5Rm7c9nEdh85PsQv4Ia23/32TGxvKpbgGIQ4DgAkOp0MfAMd/VPzNAOgPPXvD3or4sEI0xmWBo4JervPDgXVTVebovFS/B7fOqYuNeB8H7B8rQLo5UBCSGtMIk0OfygS49FmJbgYOJbO8jNEhNzyq711kSDNGCbl/xdiWA7EsoRDmgQXA0czDMBbjYMUAojEJw90/+yD7oQcBZs2EqxYptbpIp+fTosteo98ft1Tec0kAmZx4vr1Ic98IMhyHHAsB4cAAECkuzR2J7fr8y/OHvmMxRRF5XVtbdW6DBEwIf1BmPN1/+ynACCpvFOXnxBQPQhc4PxHH5yPl8SKOBYgUyvJVWf7uTWCSKbQ6ZJj3ovEOAYMnfBlX2cZwHDxLgZw52ASHMO0dbcuJQRbxXCZIsMtcUSKfBk2StPrADBSf+SsX91wufdyvlQEwaH6IwOxclLDqWu/PQUMPe8Z6uo4fzwA935bLUtfb3rUn964pHM3np/s6j6qjcZuxcVSwLQXx84lDCWGy3bbwLuxtqZGbdDiq7Pu2HMv2jPh4xQ67c4nuV085lPqdDKg6ZBCFydf+Up86tSEAos0h0klEggtembZ3Iq2tlKdUiaTSYCLTZXMVGPx4f5FABDLdBVt5yy5rN/7mgE6MUyhMxy9eqtJSw/Un52I7ADUJrUEVldBvbUPu6086+c5juUAQgwXAkj+fKXl56oHjgzUn5a0Vedj9MT1rllJ+c26fAAIMXRgleUAAn6akcpSz3wymRTcnoGhCVqhqzDIQsFAgAWOWw3QjEwmhlCQDrDAsaurNKOUiXUNTaaJ0+drpet1JoWICXgnJ4Llvdcqkj8yhg7QHEDQ7/NKMQAAjvYMfT66nltZbRCBSCZTwOTU9R51uZTze0bc3iAA5g8EQYqxQXrWe6VHXJkvFQU9Pe5VXHtODQDbGJbUfijIcADcOhMCWf6Z3qv+spPn67ukd9oMYpGpqUH7SXd9LXbulEEKzLJnxA1nbl0yiCEUXAywHHDrgcWgNl8qAgAmuEgHAULBwHJQoZSKGNpPMwChYHafW2eaNdGjX5rIzWszfXarXkVRlMZc1T6+FHm2MF5FxW6MPnBIcbOr2aqhKJXePry26aqJ31I1vLbpsm9Jh8Ph8ObSeHuVWaOiKEpjtDn6ZtZeqXJl0EYRW6FUeqvdOb0SsWJjbpC3WGOs6nQtDNdQROQMujE33F5j1WsoiqA0RptjcG4jHA7vxLAYyZlrw9GBiJ6INxaGm21GDUVRKo25qplvIzzXbozZGwmBPeNIGHvn0uZ0LI3iGiKE5l3ezyF+GJDmEEKDNIcQGqQ5hNAgzSGEBmkOITRIcwihQZpDCA3SHEJokOYQQoM0l4LQ4lB9WYFcLi/u8P7Apix2FZMkScZfNE2AmYg4ox4eir36GZqolZMlXekdKbct8IZE3TuLYw5xr4A09yqLV06fnwzqmq7euFSp/oFtyW+4OdZ6KLXzn9hyaezGp0nPL995AAAIRklEQVRv4MmKyitNmV7H2rbAm2Fo+93di0UZX/LK+neZdk/QN09j2tZLZype623Ht4xYadDJMPCnyhPJdLp8CTaaeEl36lriy6nzHYWfjBju/PmaKU2Bt45YppNJMVhPXwLNc6/AhTgAXPz9+rK/y2S95jJ4g6aA7i+Tf/SZj+Pun/0pSZZE/FaD3p76kkK5XF5QfKwl4jQac/2U104wfFqeF/1vcZJTaCpCtKe/8VjUG/Vwy8higmVBb09tSYFcnldY1ji0uNVXJrQ80nK4ME8uLyip7fIGEjwqPI15/A9j8Nu1+Y5C8pOBVWBHj5P8Dw4kFYg0lcrDN4N36msN7BZ29vLj3y3pvUFTs7G04Go3E5S1c2aB91vdmGk3U5S+yjkeqUHvmN4Ih8ObK4NVKsLYPr22GQ6Hl/qsFBF1HtxcGa5REcb26Y10b3tGPPr0VZ3DrulpV5/dTBEauyvip7fUZ1MRKqujb9w13tds1RAEoWmeidw33awnKH1V5/D4+HBnjZEiElxVN1ZmnHEPyc21pblhu4agqvqiPrhbC2Ty8E3rnbqDgXXZVYSxPaVLcTgcDme75tJ6g6Znpc9KxV+gXemzUglOsEtOKxX1DF0btFFR39JBm8paZVNFnDkXnGZKH1VJajZczdaawegnuDnt0BAax3Q4HA6H59r1iS7PG+M1qpjm1oartriMLnSat7hHb47XbPHKnWvXE6pIva8WyOzhm9o7NdJUpoHdRnPZvram9QbdIYzX4+dyiyxa4H06ZfkKnAv4AwAA0iKTGuY9vhAEvVN+qeFUhXbd4/YDBP3eAKYzaTNVLC799e9uVSsjKZFEKgaWCwEA0PPz65i2whQ9wYilCT/p4/f6OYmpIuaTK5bhb7Dx3N7D91Xv1MgS+gYDm/Xn1nTeoDuEZViA1S+OfPBFwkUJww+u0mSQXXF7Awzj9uFFZ4pMMm3HgHsxZPD6MMNlwzY+Xcz80JXrI55FOshyABwLwD+YYdZZDtSSlL+kFGJYFqSyt/TDSa/h4RtX1esPbJZrjs7kDboTcDEOoPj05tWj8Q8h5pUN+aYiycC8Z5KexwzXtCKZxKLumPRMcj5OXZfRmxlguef4ke6gqfXXtywKsQgCPcePR555YDgGwLDcq45qAIBhANz6LibqTLyBh++bDGx2a473Bq1uayvVAQCEghzLwa4mCbHWoMau06xMV5rqm6816bDJ0QEaDJeKRABKk0XR7R7iaEV5UeaHe0Gf1w/aptYzJiUAALPIxo59snwFznk9XuZoqRgAgAnGNSZSqNWYe967DDp+WWboHfSIC6XwtATgPXxHZ93zIQP/8xC8h2/d9k+M32hgs1tzGbxB0wxriFkOBDgO2ECAd2xVVp8rH63tON7INFXki0L04uSIz3DtVsQ7XmQwabnJ+6yliV9JlSaTrPu6X1F3WZm6/ii4TCbhRkev9EstYsbnHp3ysgDBwCID+eKiuqMK98D5Wul6dREe9Axcd7OABzxT8zKTTnm02nT97JWT9dBQLgPf5PUBP3AKr8ejLTfliyMupEwwsMwolGIRgDRXhrHzowMTXL7aYlJyWwtk8PBN653Kp9MPLB5cpBmOL6hO7TSe6WyVBaT3Bk1J3J+TSHBsXZl2RpxgVXqr3elaSXgEsjJooxLdROfa9el+bympqRlnjVlDUZTGbHfOzPXZKIJQ2V2RWl3tVUbeD7aqfbDdSvGOo3N8lyIuuSq91THotKsixm6Epx2qmPG2iDvvxkynTU9RlMbWt5SyQGoP3/TeqRFP4XQDm+hSHH+ysgXk34oQmmx/VoL48YE0hxAapDmE0CDNIYQGaQ4hNEhzCKFBmkMIDdIcQmiQ5hBCgzSHEBqkudck6g+R1+j5oU15WzAjx6L+ssHtS++IlKOU3ZrzNMbit8rL+pdDESfkxOiohSXHOkYWdx+mMr/hxthFyxvHztwBMW+f4hZvzM7l/rJYVwrekuzF5b8eu1G5s4gVOyS/6c7vr5Ynj9K27z/8XRPxJNE7xmM+KJ1WijA3j8/MzMzMzEzzAde2cczhWXKaqa2vpLzibPD9sLm2MFij4V/UWIhfXBp36AnK5lxYyezfEQ6nCE2WjpU+W2LQsLfBK6OU3fMciGUKhRRAJFXwb3KJZQoZxkfcMBgMBoOp4tS1O21F3P3rA2l/6uAHRyTNV0gwAAwC1zuiPxMhkioVUhGAVKFIE77iR0uWa24nSLUmBdA0HYxsZyJ+qaGp+mjy2MijnjL5x91+jrt/9t/IvPqpxPuD3v76kgK5XF5QUt8/zyTlvOI6GvV+LenyTHVFnVjrhxa3dQ/FLXXVMt/nHRNpN1ux8K15hWX1PZ5IudShZpn5oUbetsLDjcmtM96e+rh77Ra/21TOsNtlRYnsbbJ8bY3Fb42tSWuDNmprfBLez695JhzeXHBaE9zrNmfajQRh7pzbDG+uLS24mo0Rv9foYjbXricIlcZqdw6Pjw93VumJBN+8tK6jm2sLwzUaglIZa5yuuYWFufFmM0Xo4x5/KZhp1hMqx/TatENPRfxrIw5+dlc0Vsl0szESHnZ63Gk3UlTEiTBFqNnNBadVRWhs7cMu13C7TUNoasb55XSlz0YRFGWsah8cHx921hgT/G7TO8NmykpcWzdmmo0U8a5qbikSzXVtYdhhpKhobNO1QVu89JLTnCjP1Pu5hHeEN4ZrqJhfZ0bX0ZlmPaGKuVCHN8Zrttj4KhHNhcNr4zWaaETVLZpb6bNSlDX+Yu5CpznBG3Xrfm5tvEaV4MK9Fg8SHItaG/3iLHQaoyGEMvQoY2fjmlsZrtIQmpr/D8209eCq18DSAAAAAElFTkSuQmCC"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display < ./Picture/toto.png;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### TODO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- End biblio on how to perform ChIP-Seq analysis\n",
    "- Dedup BAM ?\n",
    "- Launch peak calling\n",
    "- Perform IDR & other controls\n",
    "- DE analysis\n",
    "- Si grosse != entre les IN => Why ?\n",
    "- Eric enlèverrait les duplicats de PCR\n",
    "- Impossible d'estimer par l'a priori la taille encombrée de l'ADN lors de l'IP, car AID pas en intéraction directe, mais avec un complexe qui lui est fixé sur ADN. D'ailleurs on va aussi IP des bouts d'ADN de ce complexe qui était en intéraction avec un autre locus d'ADN. (trucs qui, si on en avait la connaissance biologique, pourrait être vus sous IGV, genre quand AID est sur complexe en chr1:400, il y a fréquemment intéraction de ce complexe avec autre zone = chr4-8000 ou chr1-7000).\n",
    "- Tester analyse sans Input (si input pas fiable ?)\n",
    "- Dans locus IgH, tjs la pbmatique de mauvaise mapping à certains endroits (car soit répétitif, soit bcp trop de réarrangement dans les B cells activés sur locus IgH).\n",
    "- AID peut avoir des peaks sharp et des broads, ça dépend.\n",
    "- Comment tenir compte du ratio IP/IN qui n'est pas 50/50 niveau bio-mol ?\n",
    "- Si analyse DE pas assez robuste, s'arrêter avant, cad juste call les peaks, sur les 6 samples (3 MAR, 3 WT, voir MAR-pool et WT-pool), puis sortir liste de peaks MAR spé, WT spé, et ceux qui sont vus dans les 2 mais qui semblent être != en force. Puis faire batch script IGV ?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0) Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Context is B cells, focusing on SHM event and AID enzyme. <br>\n",
    "ChIP-Seq is performed to look at AID recruitment on Ig genes & on AID off-target genes. <br>\n",
    "In this study we have WT mouse & MAR-KO (MAR) mouse. <br>\n",
    "It is expected that MAR have lower AID recruitement on IgH, IgK, IgL & AID off-target.\n",
    "\n",
    "One of the issue is : B cells of interest are pretty rare & the AID enzyme events are pretty rare too. Thus we may not have enough data (compared to classical ChIP-Seq protocols).\n",
    "\n",
    "Samples are : 3 biological rep for both WT & MAR (and all have immunoP (IP) + Input (IN) (control), for a total of 12 samples)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1) Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1) Reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Raw reads were retrieved from Ophélie Martin external drive. They were sequenced @FTS (@Lionel & @Romain) on Illumina plateform (75bp single-end)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### FASTQ samples were barcoded, separated, cleaned but not merged\n",
    "Thus we need to do this task"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Get uniques IDs (should be 12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "path='/run/user/1000/gvfs/smb-share:server=ubox.ad.unilim.fr,share=biscem/BioInformatique/Responsables/Commun/CHIPAID';\n",
    "cd $path/Data/Reads;\n",
    "for f in *.gz;\n",
    "do echo ${f%%_*};\n",
    "done | sort -u > samples.txt;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Then cat the 48 files -> 12 FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "while read sample;\n",
    "do cat $sample* > $sample.fastq.gz;\n",
    "done < samples.txt;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Lastly, use \"easier\" names for samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mv MARsKO01.fastq.gz MAR_IN_1.fastq.gz;\n",
    "mv MARsKO02.fastq.gz MAR_IN_2.fastq.gz;\n",
    "mv MARsKO03.fastq.gz MAR_IN_3.fastq.gz;\n",
    "mv MARsKO1.fastq.gz MAR_IP_1.fastq.gz;\n",
    "mv MARsKO2.fastq.gz MAR_IP_2.fastq.gz;\n",
    "mv MARsKO3.fastq.gz MAR_IP_3.fastq.gz;\n",
    "mv WT01.fastq.gz WT_IN_1.fastq.gz;\n",
    "mv WT02.fastq.gz WT_IN_2.fastq.gz;\n",
    "mv WT03.fastq.gz WT_IN_3.fastq.gz;\n",
    "mv WT1.fastq.gz WT_IP_1.fastq.gz;\n",
    "mv WT2.fastq.gz WT_IP_2.fastq.gz;\n",
    "mv WT3.fastq.gz WT_IP_3.fastq.gz;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2) Mus musculus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference genome is the same as the one used in the PTCB project. It's the latest GRC (GRCm38.p6), minus entries not in the gencode M16 annotation (to avoid future fails with downstream analysis tools)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3) Blacklisted regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "wget http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4) Annotation tracks for IGV"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.4.1) mm10 + custom IgH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Manually retrive latest custom primer position @Ophelie"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- https://www.ensembl.org/Mus_musculus/Tools/Blast?db=core\n",
    "- Search against Mus_musculus\n",
    "- Search Tool : BLASTN\n",
    "- Search Sensitivity : Near match <br>\n",
    "Do this for the following primers : AS0, AS1, AS2, AS3 et S1, S2, S3, as described in fig_s3_locus_avec_primers_de_rt_et_probes.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then proceed to create custom annotation track for IGV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "# Retrieve latest gencode annotation (M16)\n",
    "url='ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M16';\n",
    "name='gencode.vM16.chr_patch_hapl_scaff';\n",
    "axel -q $url/$name.annotation.gtf.gz;\n",
    "unpigz $name.annotation.gtf.gz;\n",
    "\n",
    "# Create a file with our custom annotations\n",
    "touch custom_igh.gtf;\n",
    "echo 'chr12 . hs7 113221438 113221581 . - . Name=hs7' >> custom_igh.gtf;\n",
    "echo 'chr12 . hs6 113223561 113223688 . - . Name=hs6' >> custom_igh.gtf;\n",
    "echo 'chr12 . hs5 113224745 113224923 . - . Name=hs5' >> custom_igh.gtf;\n",
    "echo 'chr12 . hs4 113228349 113228519 . - . Name=hs4' >> custom_igh.gtf;\n",
    "echo 'chr12 . hs3b 113232345 113232538 . - . Name=hs3b' >> custom_igh.gtf;\n",
    "echo 'chr12 . hs1.2 113244593 113244972 . - . Name=hs1.2' >> custom_igh.gtf;\n",
    "echo 'chr12 . hs3a 113253847 113254040 . - . Name=hs3a' >> custom_igh.gtf;\n",
    "echo 'chr12 . HRE2 113315188 113314617 . - . Name=HRE2' >> custom_igh.gtf;\n",
    "echo 'chr12 . HRE1 113320944 113321537 . - . Name=HRE1' >> custom_igh.gtf;\n",
    "echo 'chr12 . IgD 113416247 113418359 . - . Name=IgD' >> custom_igh.gtf;\n",
    "echo 'chr12 . Mar3 113426976 113427406 . - . Name=Mar3' >> custom_igh.gtf;\n",
    "echo 'chr12 . cEµ 113427408 113427624 . - . Name=cEµ' >> custom_igh.gtf;\n",
    "echo 'chr12 . Mar5 113427624 113427970 . - . Name=Mar5' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_AS0 113429300 113429319 . - . Name=primer_AS0' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_AS1 113428766 113428785 . - . Name=primer_AS1' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_AS2 113428491 113428507 . - . Name=primer_AS2' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_AS3 113428417 113428436 . - . Name=primer_AS3' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_S1 113428372 113428393 . - . Name=primer_S1' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_S2 113427973 113427996 . - . Name=primer_S2' >> custom_igh.gtf;\n",
    "echo 'chr12 . primer_S3 113427554 113427574 . - . Name=primer_S3' >> custom_igh.gtf;\n",
    "sed -i 's/ /\\t/g' custom_igh.gtf;\n",
    "\n",
    "# Unsorted concatenation of M16 + custom\n",
    "cat $name\".annotation.gtf\" custom_igh.gtf > $name\"_and_20_custom_igh.annotation.gtf\";\n",
    "rm $name\".annotation.gtf\" custom_igh.gtf;\n",
    "\n",
    "# Sort with gff3sort (because GNU sort has a bug : Lines with the same chromosomes and start positions would be placed randomly)\n",
    "file=$name'_and_20_custom_igh.annotation.gtf';\n",
    "gff3sort.pl --precise --chr_order natural $file |bgzip > ${file%.annotation*}_gff3sort.annotation.gtf.gz;\n",
    "tabix -p gff ${file%.annotation*}_gff3sort.annotation.gtf.gz;\n",
    "rm $name'_and_20_custom_igh.annotation.gtf';"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.4.2) mm9 (to view Casellas results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Retrieve corresponding gencode annotation\n",
    "# https://www.biostars.org/p/291445/\n",
    "url='ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M1';\n",
    "name='gencode.vM1';\n",
    "axel -q $url/$name.annotation.gtf.gz;\n",
    "unpigz $name.annotation.gtf.gz;\n",
    "\n",
    "# Sort with gff3sort (because GNU sort has a bug : Lines with the same chromosomes and start positions would be placed randomly)\n",
    "gff3sort.pl --precise --chr_order natural $name.annotation.gtf |bgzip > $name\"_gff3sort.annotation.gtf.gz\";\n",
    "tabix -p gff $name\"_gff3sort.annotation.gtf.gz\";\n",
    "rm $name.annotation.gtf;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2) Quality control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### FastQC v0.11.2 @genotoul"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "#!/bin/bash\n",
    "#$ -q workq\n",
    "#$ -M erwan.scaon@unilim.fr\n",
    "#$ -m bea\n",
    "#$ -l mem=8G\n",
    "#$ -l h_vmem=10G\n",
    "#$ -N fastqc_chipaid\n",
    "#$ -o /home/escaon/work/CHIPAID/Verbose/fastqc.o\n",
    "#$ -e /home/escaon/work/CHIPAID/Verbose/fastqc.e\n",
    "\n",
    "##############################\n",
    "### Paths, tools & modules ###\n",
    "##############################\n",
    "work='/home/escaon/work/CHIPAID';\n",
    "\n",
    "##############\n",
    "### FASTQC ###\n",
    "##############\n",
    "cd $work/Data;\n",
    "for f in *.fastq.gz;\n",
    "do echo $f; fastqc --outdir $work -f fastq $f;\n",
    "done;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### MultiQC 1.4 @genotoul"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "#!/bin/bash\n",
    "#$ -q workq\n",
    "#$ -M erwan.scaon@unilim.fr\n",
    "#$ -m bea\n",
    "#$ -l mem=8G\n",
    "#$ -l h_vmem=10G\n",
    "#$ -N multiqc_chipaid\n",
    "#$ -o /home/escaon/work/CHIPAID/Verbose/multiqc.o\n",
    "#$ -e /home/escaon/work/CHIPAID/Verbose/multiqc.e\n",
    "\n",
    "##############################\n",
    "### Paths, tools & modules ###\n",
    "##############################\n",
    "work='/home/escaon/work/CHIPAID';\n",
    "fastqc_dir='/home/escaon/work/CHIPAID/QC/FastQC';\n",
    "\n",
    "###############\n",
    "### MULTIQC ###\n",
    "###############\n",
    "multiqc $fastqc_dir \\\n",
    "        -n chipaid \\\n",
    "        -o $work \\\n",
    "        -m fastqc \\\n",
    "        -f;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Thoughts on QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3 biological rep for both WT & MAR, thus 6 intial samples to look at (same initial sample is splited to provide IN & IP) :\n",
    "\n",
    "WT_1 :<br>\n",
    "WT_IP_1 has bad sequence qual & under average read len (56).\n",
    "\n",
    "WT_2 :<br>\n",
    "WT_IN_2 has under average read len (61) and mixed sequence qual.\n",
    "WT_IP_2 has a whooping 75.5M reads.\n",
    "\n",
    "WT_3 :<br>\n",
    "Seems good : ~24M reads in both IP & IN, ~75 len, good qual.\n",
    "\n",
    "MAR_1 :<br>\n",
    "MAR_IP_1 has bad seq qual and under average read len (57).\n",
    "MAR_IN_1 has mixed sequence qual.\n",
    "\n",
    "MAR_2 :<br>\n",
    "MAR_IP_2 has abberant GC%, low lenght (mean 46 vs expected 75) and few sequences.<br>\n",
    "Associated MAR_IN_2 has 190.5M sequences ! (with 75 len & good qual).<br>\n",
    "\n",
    "MAR_3 :<br>\n",
    "Seems good : ~70 len, good qual, 38.2M reads in IN and 4.5 in IP, which could be OK given the biological context (less IP expected in MAR)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3) Align reads vs reference genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1) Build index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given that reads are shorter here than in the PTCB project, let's rebuild the genome index with this in mind (sjdbOverhang parameter 150 -> 74)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### STAR 2.5.3a @CALI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "#!/bin/bash\n",
    "#SBATCH --partition=normal\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --mem-per-cpu=6144\n",
    "#SBATCH --time=1-23:59:59\n",
    "#SBATCH --mail-user=erwan.scaon@unilim.fr\n",
    "#SBATCH --mail-type=begin\n",
    "#SBATCH --mail-type=end\n",
    "#SBATCH --error=/home/scaonp01/scratch/PTCB/Verbose/star_idx_custom_chipaid.e\n",
    "#SBATCH --output=/home/scaonp01/scratch/PTCB/Verbose/star_idx_custom_chipaid.o\n",
    "#SBATCH --job-name=star_idx_custom\n",
    "\n",
    "##############################\n",
    "### Paths, tools & modules ###\n",
    "##############################\n",
    "nt=8;\n",
    "export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK;\n",
    "star='/home/scaonp01/Software/STAR_2.5.3a/bin/Linux_x86_64';\n",
    "ptcb='/home/scaonp01/PTCB/Data/Mus_musculus';\n",
    "scratch='/home/scaonp01/scratch/PTCB';\n",
    "\n",
    "############\n",
    "### STAR ###\n",
    "############\n",
    "mkdir -p $ptcb/GRCm38.p6_custom_chipaid_STAR;\n",
    "cd $ptcb/GRCm38.p6_custom_chipaid_STAR;\n",
    "$star/STAR --runThreadN $nt \\\n",
    "           --runMode genomeGenerate \\\n",
    "           --genomeDir $ptcb/GRCm38.p6_custom_chipaid_STAR \\\n",
    "           --genomeFastaFiles $ptcb/GRCm38.p6_custom_STAR/GCA_000001635.8_GRCm38.p6_genomic_renamed_subsampled.fna \\\n",
    "           --sjdbGTFfile $ptcb/GRCm38.p6_custom_STAR/gencode.vM16.chr_patch_hapl_scaff.annotation.gtf \\\n",
    "           --sjdbOverhang 74;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2) Alignments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### STAR 2.5.3a @CALI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initial script : star_WT_IN_2.sh <br>\n",
    "Nb_1 : We use gapped aligner based on biostar handbook advise. <br>\n",
    "Nb_2 : Default parameters (Did not find any guidelines for ChIP-Seq such as Encode alignement parameters advise for RNA-Seq). <br>\n",
    "Nb_3 : We keep unaligned reads in a separated FASTQ just in case."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "#!/bin/bash\n",
    "#SBATCH --partition=normal\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --mem-per-cpu=6144\n",
    "#SBATCH --time=1-23:59:59\n",
    "#SBATCH --mail-user=erwan.scaon@unilim.fr\n",
    "#SBATCH --mail-type=BEGIN\n",
    "#SBATCH --mail-type=END\n",
    "#SBATCH --error=/home/scaonp01/scratch/CHIPAID/Verbose/star_WT_IN_2.e\n",
    "#SBATCH --output=/home/scaonp01/scratch/CHIPAID/Verbose/star_WT_IN_2.o\n",
    "#SBATCH --job-name=star_WT_IN_2\n",
    "\n",
    "##############################\n",
    "### Paths, tools & modules ###\n",
    "##############################\n",
    "nt=8;\n",
    "export OMP_NUM_THREADS=$SLURM_CPUS_PER_TASK;\n",
    "star='/home/scaonp01/Software/STAR_2.5.3a/bin/Linux_x86_64/STAR';\n",
    "samtools='/home/scaonp01/Software/Samtools_1.6/Install/bin/samtools';\n",
    "save='/home/scaonp01/CHIPAID';\n",
    "work='/home/scaonp01/scratch/CHIPAID';\n",
    "\n",
    "############\n",
    "### STAR ###\n",
    "############\n",
    "cd $save/Data/Reads;\n",
    "for id in 'WT_IN_2';\n",
    "do $star --genomeDir $save/Data/GRCm38.p6_custom_chipaid_STAR \\\n",
    "         --readFilesIn $id.fastq.gz \\\n",
    "         --outFileNamePrefix $work/$id\"_vs_GRCm38.p6_custom_\" \\\n",
    "         --runThreadN $nt \\\n",
    "         --readFilesCommand zcat \\\n",
    "         --outSAMtype BAM SortedByCoordinate \\\n",
    "         --outReadsUnmapped Fastx;\n",
    "done;\n",
    "\n",
    "######################\n",
    "### Compress Stuff ###\n",
    "######################\n",
    "cd $work;\n",
    "mv $id\"_vs_GRCm38.p6_custom_Unmapped.out.mate1\" $id\"_vs_GRCm38.p6_custom_Unmapped.out.mate1.fastq\";\n",
    "gzip $id\"_vs_GRCm38.p6_custom_Unmapped.out.mate1.fastq\";\n",
    "$samtools index $id\"_vs_GRCm38.p6_custom_Aligned.sortedByCoord.out.bam\";"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Duplicate script for the 11 other samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "for id in 'MAR_IN_1' 'MAR_IN_2' 'MAR_IN_3' 'MAR_IP_1' 'MAR_IP_2' 'MAR_IP_3' 'WT_IN_1' 'WT_IN_3' 'WT_IP_1' 'WT_IP_2' 'WT_IP_3';\n",
    "do cp star_WT_IN_2.sh star_$id.sh;\n",
    "   sed -i \"s/WT_IN_2/$id/\" star_$id.sh;\n",
    "   sbatch star_$id.sh;\n",
    "done;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3) Alignements QC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### MultiQC 1.4 @local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "path='/run/user/1000/gvfs/smb-share:server=ubox.ad.unilim.fr,share=biscem/BioInformatique/Responsables/Commun/CHIPAID';\n",
    "cd $path/Output/1_STAR/QC;\n",
    "multiqc ./ -n star_vs_custom -o ./ -m star -f > ./multiqc.verbose 2>&1;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## 4) Number of alignments per region of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the potential low throughput of this ChIP-Seq experiment, we want to check if we have enough alignments on known regions of interest.  <br>\n",
    "\n",
    "Those regions are IgH, IgK & IgL loci : <br>\n",
    "https://www.ncbi.nlm.nih.gov/gene/111507 <br>\n",
    "https://www.ncbi.nlm.nih.gov/gene/243469 <br>\n",
    "https://www.ncbi.nlm.nih.gov/gene/111519 <br>\n",
    "chr6:67555636-70726754\tIgK <br>\n",
    "chr12:113258768-116009954\tIgH <br>\n",
    "chr16:19026858-19260844 IgL <br>\n",
    "\n",
    "And genes of interest (105) in the context of MAR-KO mutant, previously created for the PTCB project (genes_of_interest_mar.tsv) : <br>\n",
    "All members of the ARID family <br>\n",
    "All members of thee SATB family <br>\n",
    "Cux1 <br>\n",
    "AID off targets genes in Liu et al., Nature Letters, 02/2008, Supp Fig1 <br>\n",
    "Genes on the first page of Yamane et al., Nature Immunology, 01/2011, S1 Table <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1) Get \"chr\", \"start\", \"end\" and \"gene_name\" for each of the 105 gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "R;\n",
    "library('biomaRt')\n",
    "mar_goi=read.table(\"./genes_of_interest_mar.tsv\")\n",
    "colnames(mar_goi)[1] <- 'ensembl_gene_id'\n",
    "\n",
    "mmusculus = useMart(\"ENSEMBL_MART_ENSEMBL\",\n",
    "                    dataset=\"mmusculus_gene_ensembl\")\n",
    "\n",
    "mm_infos <- getBM(attributes=c('ensembl_gene_id',\n",
    "                               'chromosome_name',\n",
    "                               'start_position',\n",
    "                               'end_position',\n",
    "                               'external_gene_name'),\n",
    "                  mart = mmusculus)\n",
    "\n",
    "mar_bed <- merge(x = mar_goi,\n",
    "                 y = mm_infos,\n",
    "                 by = \"ensembl_gene_id\",\n",
    "                 all.x = TRUE)\n",
    "\n",
    "write.table(mar_bed,\n",
    "            file='./mar_goi_2.tsv',\n",
    "            row.names=FALSE,\n",
    "            col.names=FALSE,\n",
    "            sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Check coherence between biomart (ensembl) & GRC seqs names (in BAM / ref genome) <br>\n",
    "In mar_goi.tsv : \"1\" ... \"19\" \"X\" <br>\n",
    "In bam : >chr1, >GL456210.1, >GL456211.1 <br>\n",
    "\n",
    "Thus it's okay, nothing special in goi, just need to add \"chr\"."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2) Create BED for bedtools multicov"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Add chr & move col 1 to col 4 (bed format is chr, start, end, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "awk -F '\\t' -vOFS='\\t' '{print $2, $3, $4, $5}' mar_goi_2.tsv > temp_1.tsv;\n",
    "awk -F '\\t' -vOFS='\\t' '{$1 = \"chr\"$1}1' temp_1.tsv > temp_2.tsv;\n",
    "sed -i 's/\\\"//g' temp_2.tsv;\n",
    "# Add IgH, IgK and IgL loci manually\n",
    "# And sort BED, otherwise we will get error with downstream tools\n",
    "sort -k2,2n -k3,3n temp_2.tsv > temp_3.tsv;\n",
    "sort -V -k1 temp_3.tsv > mar_goi_2.bed;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3) Get number of alignments per regions in our BED"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### bedtools v2.25.0 @local"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://bedtools.readthedocs.io/en/latest/content/tools/multicov.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### CODE ###\n",
    "# bamlist <=> ls -1 *.bam in BAM folder @CALI\n",
    "bedtools multicov \\\n",
    "         -bams `cat bamlist.txt` \\\n",
    "         -bed mar_goi_2_igx.bed \\\n",
    "         > alignments_per_region.tsv;\n",
    "sed -i '1 i\\chr\\tstart\\tend\\tgene\\tMAR_IN_1\\tMAR_IN_2\\tMAR_IN_3\\tMAR_IP_1\\tMAR_IP_2\\tMAR_IP_3\\tWT_IN_1\\tWT_IN_2\\tWT_IN_3\\tWT_IP_1\\tWT_IP_2\\tWT_IP_3' alignments_per_region.tsv;"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
